### ISSUE TITLE
Memory store's `put()` operation is prohibitively slow for loading large cache

### ISSUE THREAD

**Issue by kriszyp** [SitePen/dstore#180]
If you use the `Cache` module with the typical/default configuration, a master Rest store and a Memory store for the cache, and fetch data, the data is loaded into the Memory store using repeated `put()` for each loaded entry. This requires constructing a `Deferred` object and resolving it, which is rather expensive. In our situation, loading 45,000 entries, requires 30 seconds to process (inside the Cache processing).

Better performance is pretty trivial to achieve: in the cache load loop, we can simply call `putSync`, if it is available, which appears to be about 40x faster.

**Comment by rainmanhhh** [comment]
maybe a workaround is using a simulate Deferred object in Memory.put instead of constructing new instances?

**Comment by dylans** [comment]
I'm curious about a few things:

- Would it be this slow if put relied on native promises? I'm curious to know the performance of native promises vs. sync code
- At the point of caching, are we guaranteed that all items are in memory, or will some items still be promises yet to resolve (I'm assuming @kriszyp's scenario everything is in memory)?

**Comment by kriszyp** [comment]
> Would it be this slow if put relied on native promises?
I am haven't verified this in this case, but I am pretty sure that native promises are plenty fast enough (faster than the internals of Memory's `putSync`). IIRC, I have tested native promises performance before, and you can generally create around a million per second. Using native promises would be a good solution, just a little more work :).

> At the point of caching, are we guaranteed that all items are in memory, or will some items still be promises yet to resolve?

The cache is loaded after items are loaded, specifically this is where the repeated `put()` calls occur (after the fetch results are loaded): https://github.com/SitePen/dstore/blob/1.1/Cache.js#L84

**Comment by kfranqueiro** [comment]
Am I right in thinking that native promises would technically break backcompat because dojo/Deferred does not follow A+ and will resolve synchronously?  (Not to mention any other details differentiating dojo/Deferred promises from native ones...)

Admittedly 45k seems like an awful lot of records to be caching at once client-side, but I am pretty sure we do already have a check somewhere (possibly in Cache) that optimizes for the caching store being a Memory store, so having another Memory-specific optimization to prefer `putSync` & co. would not be a big deal to me.

**Comment by kfranqueiro** [comment]
If you want to PR your fix Kris, I'd certainly be up for including it in a new 1.1.x release.  I'm still trying to figure out where I thought I remembered us making a similar optimization/concession like this before for Memory stores though...

**Comment by kfranqueiro** [comment]
Oh.  Actually it looks like this was already committed, at least to the 1.1 branch.  I'll have to cherry-pick it forward into 1.x (which effectively acts like master did before we started thinking about converting to typescript).

**Comment by kriszyp** [comment]
Oh, OK, thanks for the help. No rush on a release, it was pretty easy to workaround in our code (`this.cachingStore.put = this.cachingStore.putSync`) :).
